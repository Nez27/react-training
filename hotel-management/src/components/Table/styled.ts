import styled from 'styled-components';

// Interfaces
import { ITable } from '../../globals/interfaces';

const StyledTable = styled.div`
  border: 1px solid var(--border-color);

  font-size: 20px;
  border-radius: var(--radius-md);
`;

const CommonRow = styled.div<ITable>`
  display: grid;
  grid-template-columns: ${(props) => props.columns};
  column-gap: 10px;
  align-items: center;
  justify-items: center;
`;

const StyledBody = styled.div`
  font-size: var(--fs-sm-x);
`;

const StyledRow = styled(CommonRow)`
  padding: 20px;

  &:not(:last-child) {
    border-bottom: 1px solid var(--border-color);
  }
`;

const StyledHeader = styled(CommonRow)`
  padding: 10px 20px;

  border-bottom: 1px solid var(--border-color);
  background-color: var(--header-table-color);
  border-top-left-radius: var(--radius-md);
  border-top-right-radius: var(--radius-md);

  font-size: var(--fs-sm);
  text-transform: capitalize;
  font-weight: 600;
`;

export { StyledBody, StyledHeader, StyledRow, StyledTable };
