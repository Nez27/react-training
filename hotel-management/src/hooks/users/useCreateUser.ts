import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { createUser as createUserFn } from '@service/userServices';

// Constants
import { ADD_SUCCESS } from '@constant/messages';

// Hooks
import { useUserRoomAvailable } from '@hook/useUserRoomAvailable';

/**
 * Create user in database
 * @returns The status of creating users and createUser function
 */
const useCreateUser = () => {
  const queryClient = useQueryClient();
  const { dispatch } = useUserRoomAvailable();

  const { mutate: createUser, isPending: isCreating } = useMutation({
    mutationFn: createUserFn,
    onSuccess: (user) => {
      toast.success(ADD_SUCCESS);
      queryClient.invalidateQueries({ queryKey: ['users'] });

      // Add user to global statement
      dispatch!({
        type: 'addUser',
        payload: [{ id: user.id, name: user.name }],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isCreating, createUser };
};

export { useCreateUser };
