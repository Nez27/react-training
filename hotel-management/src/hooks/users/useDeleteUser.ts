import { useMutation, useQueryClient } from '@tanstack/react-query';
import toast from 'react-hot-toast';

// Services
import { deleteUser as deleteUserFn } from '@service/userServices';

// Constants
import * as messages from '@constant/messages';

/**
 * Delete the user from database
 * @returns The boolean of status deleting and deleteUser function
 */
const useDeleteUser = () => {
  const queryClient = useQueryClient();

  const {
    isPending: isDeleting,
    mutate: deleteUser,
    isSuccess
  } = useMutation({
    mutationFn: deleteUserFn,
    onSuccess: () => {
      toast.success(messages.DELETE_SUCCESS);
      queryClient.invalidateQueries({
        queryKey: ['users'],
      });
    },
    onError: (err) => toast.error(err.message),
  });

  return { isDeleting, deleteUser, isSuccess };
};

export { useDeleteUser };
