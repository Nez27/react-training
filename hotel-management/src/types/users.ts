interface IUser {
  id: number;
  name: string;
  phone: string;
}

export type { IUser };
