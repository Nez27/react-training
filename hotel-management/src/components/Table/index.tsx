import { ReactNode, useContext } from 'react';

// Components
import { StyledBody, StyledHeader, StyledRow, StyledTable } from './styled';

// Interfaces
import { ITable, ITableBody } from '../../globals/interfaces';

// Contexts
import TableContext from '../../contexts/TableContext';

type CallbackMapFunc<T> = (value: T, index: number, array: T[]) => ReactNode;

const Table = ({ columns, children }: ITable) => {
  return (
    <TableContext.Provider value={columns!}>
      <StyledTable>{children}</StyledTable>
    </TableContext.Provider>
  );
};

const Header = ({ children }: ITable) => {
  const columns = useContext(TableContext);
  return <StyledHeader columns={columns}>{children}</StyledHeader>;
};

const Body = <T,>({ data, render }: ITableBody<T>) => {
  data!.length > 0 && (
    <StyledBody>{data?.map(render as CallbackMapFunc<T>)}</StyledBody>
  );
};

const Row = ({ children }: ITable) => {
  const columns = useContext(TableContext);
  return <StyledRow columns={columns}>{children}</StyledRow>;
};

Table.Header = Header;
Table.Body = Body;
Table.Row = Row;

export default Table;
