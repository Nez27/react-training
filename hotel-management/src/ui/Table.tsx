import React, { createContext, useContext } from 'react';
import styled from 'styled-components';

const TableContext = createContext('');

const StyledTable = styled.div`
  border: 1px solid var(--border-color);

  font-size: 20px;
  border-radius: var(--radius-sm);
  overflow: hidden;
`;

const CommonRow = styled.div<ITable>`
  display: grid;
  grid-template-columns: ${(props) => props.columns};
  column-gap: 10px;
  align-items: center;
`;

const StyledEmpty = styled.p`
  font-size: var(--fs-sm);
  text-align: center;
  padding: 20px;
`;

const StyledBody = styled.div`
  margin: 10px;
`;

const StyledRow = styled(CommonRow)`
  padding: 10px;
`;

interface ITable {
  columns?: string;
  children?: React.ReactNode;
}

interface ITableBody<T> {
  data?: T[];
  render?: (value: T, index: number, array: T[]) => T;
}

const Table = ({ columns = '', children }: ITable) => {
  return (
    <TableContext.Provider value={columns}>
      <StyledTable>{children}</StyledTable>
    </TableContext.Provider>
  );
};

const StyledHeader = styled(CommonRow)`
  font-size: var(--fs-sm);
  padding: 10px 20px;
  border-bottom: 1px solid var(--border-color);
  text-transform: capitalize;
  font-weight: 600;
`;

const Header = ({ children }: ITable) => {
  const columns = useContext(TableContext);
  return <StyledHeader columns={columns}>{children}</StyledHeader>;
};

const Body = <T extends string>({ data, render }: ITableBody<T>) => {
  if (data && !data.length) return <StyledEmpty>No data!</StyledEmpty>;

  return <StyledBody>{data?.map(render!)}</StyledBody>;
};

const Row = ({ children }: ITable) => {
  const columns = useContext(TableContext);
  return <StyledRow columns={columns}>{children}</StyledRow>;
};

Table.Header = Header;
Table.Body = Body;
Table.Row = Row;

export default Table;
