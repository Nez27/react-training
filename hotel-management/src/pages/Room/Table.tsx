import { useSearchParams } from 'react-router-dom';
import { useEffect, useState } from 'react';
import toast from 'react-hot-toast';

// Components
import { HiSquare2Stack } from 'react-icons/hi2';
import { HiTrash } from 'react-icons/hi';
import { StyledOperationTable } from './styled';
import Menus from '../../components/Menus';
import Table from '../../components/Table';
import Direction from '../../commons/styles/Direction';
import Message from '../../components/Message';
import Search from '../../components/Search';
import SortBy from '../../components/SortBy';
import OrderBy from '../../components/OrderBy';

// Types
import { TRoom } from '../../globals/types';

// Constants
import { useFetch } from '../../hooks/useFetch';
import { ROOM_PATH } from '../../constants/path';
import { STATUS_CODE } from '../../constants/statusCode';
import { CONFIRM_DELETE, DELETE_SUCCESS } from '../../constants/messages';
import { ROOM_PAGE } from '../../constants/variables';

// Styled
import Spinner from '../../commons/styles/Spinner';

// Utils
import { sendRequest } from '../../helpers/sendRequest';

interface IRoomRow {
  room: TRoom;
  openFormDialog: () => void;
  setRoom: React.Dispatch<React.SetStateAction<TRoom | null>>;
  reload: boolean;
  setReload: React.Dispatch<React.SetStateAction<boolean>>;
}

const RoomRow = ({
  room,
  openFormDialog,
  setRoom,
  reload,
  setReload,
}: IRoomRow) => {
  const handleOnEdit = (room: TRoom) => {
    setRoom(room);
    openFormDialog();
  };

  const handleOnDelete = async (room: TRoom) => {
    if (confirm(CONFIRM_DELETE)) {
      const response = await sendRequest(
        ROOM_PATH + `/${room.id}`,
        null,
        'DELETE',
      );

      if (response.statusCode === STATUS_CODE.OK) {
        toast.success(DELETE_SUCCESS);

        // Reload table
        setReload(!reload);
      }
    }
  };

  const { id, name, price, status } = room;

  // prettier-ignore
  const statusText = status
    ? 'Unavailable' 
    : 'Available';

  return (
    <Table.Row>
      <div>{id}</div>
      <div>{name}</div>
      <div>{price}</div>
      <div>{statusText}</div>

      <Menus.Menu>
        <Menus.Toggle id={id.toString()} />

        <Menus.List id={id.toString()}>
          <Menus.Button
            icon={<HiSquare2Stack />}
            onClick={() => handleOnEdit(room)}
          >
            Edit
          </Menus.Button>
          <Menus.Button icon={<HiTrash />} onClick={() => handleOnDelete(room)}>
            Delete
          </Menus.Button>
        </Menus.List>
      </Menus.Menu>
    </Table.Row>
  );
};

interface IRoomTable {
  reload: boolean;
  setReload: React.Dispatch<React.SetStateAction<boolean>>;
  openFormDialog: () => void;
  setRoom?: React.Dispatch<React.SetStateAction<TRoom | null>>;
}

const RoomTable = ({
  reload,
  setReload,
  openFormDialog,
  setRoom,
}: IRoomTable) => {
  const [nameSearch, setNameSearch] = useState('');
  const [searchParams] = useSearchParams();
  const [rooms, setRooms] = useState<TRoom[]>([]);
  const sortByValue = searchParams.get('sortBy')
    ? searchParams.get('sortBy')!
    : '';
  const orderByValue = searchParams.get('orderBy')
    ? searchParams.get('orderBy')!
    : '';

  const { data, isPending, errorFetchMsg } = useFetch(
    'rooms',
    'name',
    nameSearch,
    sortByValue,
    orderByValue,
    reload,
  );

  useEffect(() => {
    if (data) {
      setRooms(data);
    } else {
      setRooms([]);
    }

    if (errorFetchMsg) {
      console.error(errorFetchMsg);
    }
  }, [data, errorFetchMsg]);

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ROOM_PAGE.ORDERBY_OPTIONS} />

          <SortBy options={ROOM_PAGE.SORTBY_OPTIONS} />
          <Search
            setValueSearch={setNameSearch}
            setPlaceHolder="Search by name..."
          />
        </StyledOperationTable>

        {isPending && <Spinner />}

        {rooms.length ? (
          <Menus>
            <Table columns="10% 40% 20% 20% 5%">
              <Table.Header>
                <div>Id</div>
                <div>Name</div>
                <div>Price</div>
                <div>Status</div>
              </Table.Header>
              <Table.Body<TRoom>
                data={rooms}
                render={(room: TRoom) => (
                  <RoomRow
                    room={room}
                    key={room.id}
                    reload={reload}
                    setReload={setReload}
                    openFormDialog={openFormDialog}
                    setRoom={setRoom!}
                  />
                )}
              />
            </Table>
          </Menus>
        ) : (
          !isPending && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default RoomTable;
