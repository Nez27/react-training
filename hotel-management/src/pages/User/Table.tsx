import { useSearchParams } from 'react-router-dom';
import { Dispatch, SetStateAction, useEffect, useState } from 'react';
import toast from 'react-hot-toast';

// Components
import { RiEditBoxFill } from 'react-icons/ri';
import { IoExit } from 'react-icons/io5';
import Menus from '../../components/Menus';
import Table from '../../components/Table';
import Direction from '../../commons/styles/Direction';
import Message from '../../components/Message';
import Search from '../../components/Search';
import SortBy from '../../components/SortBy';
import OrderBy from '../../components/OrderBy';

// Types
import { Nullable } from '../../types/common';
import { TUser } from '../../types/user';

// Hooks
import { useFetch } from '../../hooks/useFetch';

// Constants
import { STATUS_CODE } from '../../constants/responseStatus';
import { ORDERBY_OPTIONS, USER_PAGE } from '../../constants/variables';
import { CONFIRM_MESSAGE } from '../../constants/messages';

// Styled
import { StyledOperationTable } from './styled';
import Spinner from '../../commons/styles/Spinner';

// Services
import { updateRoomStatus } from '../../services/roomServices';
import { checkOutUser } from '../../services/userServices';

interface IUserRow {
  user: TUser;
  openFormDialog: () => void;
  setUser: Dispatch<SetStateAction<Nullable<TUser>>>;
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
}

const UserRow = ({
  user,
  openFormDialog,
  setUser,
  reload,
  setReload,
}: IUserRow) => {
  const handleEdit = (user: TUser) => {
    setUser(user);
    openFormDialog();
  };

  const handleCheckOut = async (user: TUser) => {
    if (confirm(CONFIRM_MESSAGE)) {
      const resUpdateStatus = await updateRoomStatus(user.roomId, false);
      const resCheckoutUser = await checkOutUser(user);

      if (
        resCheckoutUser?.statusCode === STATUS_CODE.OK &&
        resUpdateStatus?.statusCode === STATUS_CODE.OK
      ) {
        toast.success('Check out complete!');

        // Reload table
        setReload(!reload);
      }
    }
  };

  const { id, name, identifiedCode, phone, roomId } = user;

  return (
    <Table.Row>
      <div>{id}</div>
      <div>{name}</div>
      <div>{identifiedCode}</div>
      <div>{phone}</div>
      <div>{
        roomId 
          ? roomId 
          : 'None'
      }</div>

      <Menus.Menu>
        <Menus.Toggle id={id.toString()} />

        <Menus.List id={id.toString()}>
          <Menus.Button
            icon={<RiEditBoxFill />}
            onClick={() => handleEdit(user)}
          >
            Edit
          </Menus.Button>
          <Menus.Button icon={<IoExit />} onClick={() => handleCheckOut(user)}>
            Check out
          </Menus.Button>
        </Menus.List>
      </Menus.Menu>
    </Table.Row>
  );
};

interface IUserTable {
  reload: boolean;
  setReload: Dispatch<SetStateAction<boolean>>;
  openFormDialog: () => void;
  setUser?: Dispatch<SetStateAction<Nullable<TUser>>>;
}

const UserTable = ({
  reload,
  setReload,
  openFormDialog,
  setUser,
}: IUserTable) => {
  const [users, setUsers] = useState<TUser[]>([]);
  const [phoneSearch, setPhoneSearch] = useState('');
  const [searchParams] = useSearchParams();
  const sortByValue = searchParams.get('sortBy')
    ? searchParams.get('sortBy')!
    : '';
  const orderByValue = searchParams.get('orderBy')
    ? searchParams.get('orderBy')!
    : '';

  const { data, isPending, errorFetchMsg } = useFetch(
    'users',
    'phone',
    phoneSearch,
    sortByValue,
    orderByValue,
    reload
  );

  useEffect(() => {
    if (data) {
      setUsers(data);
    } else {
      setUsers([]);
    }

    if (errorFetchMsg) {
      console.error(errorFetchMsg);
    }
  }, [data, errorFetchMsg, setUsers]);

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={USER_PAGE.SORTBY_OPTIONS} />
          <Search
            setValueSearch={setPhoneSearch}
            setPlaceHolder="Search by phone..."
          />
        </StyledOperationTable>

        {isPending && <Spinner />}

        {users.length ? (
          <Menus>
            <Table columns="10% 30% 20% 20% 10% 5%">
              <Table.Header>
                <div>Id</div>
                <div>Name</div>
                <div>Identified Code</div>
                <div>Phone</div>
                <div>Room Id</div>
              </Table.Header>
              <Table.Body<TUser>
                data={users}
                render={(user: TUser) => (
                  <UserRow
                    user={user}
                    key={user.id}
                    reload={reload}
                    setReload={setReload}
                    openFormDialog={openFormDialog}
                    setUser={setUser!}
                  />
                )}
              />
            </Table>
          </Menus>
        ) : (
          !isPending && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default UserTable;
