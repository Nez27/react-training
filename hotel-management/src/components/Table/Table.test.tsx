import renderer from 'react-test-renderer';

// Components
import Table from '.';

// Types
import { IUser } from '@type/users';

interface ITableRow {
  user: IUser;
}

describe('Table', () => {
  const columnName = ['Column 1', 'Column 2'];
  const tempUser: IUser[] = [
    {
      id: 1,
      name: 'Nezumi',
      phone: '123',
      isBooked: true,
    },
    {
      id: 2,
      name: 'Loi Phan',
      phone: '453',
      isBooked: false,
    },
  ];
  const TableRow = ({ user }: ITableRow) => {
    const { id, name, phone, isBooked } = user;

    return (
      <Table.Row>
        <div>{id}</div>
        <div>{name}</div>
        <div>{phone}</div>
        <div>{isBooked}</div>
      </Table.Row>
    );
  };
  const renderRow = (user: IUser) => <TableRow user={user} key={user.id} />;

  const wrapper = renderer.create(
    <Table columns="10% 35% 30% 15% 10%">
      <Table.Header headerColumn={columnName} />
      <Table.Body<IUser> data={tempUser} render={renderRow} />
    </Table>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
