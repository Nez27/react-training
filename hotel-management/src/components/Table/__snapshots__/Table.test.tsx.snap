// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table testing snapshot render 1`] = `
<div
  className="sc-bdnyFh eEibLN"
>
  <div
    className="sc-gtsqUy sc-eCAqax eucOOU hcIWQg"
    columns="10% 35% 30% 15% 10%"
  >
    <div>
      Column 1
    </div>
    <div>
      Column 2
    </div>
  </div>
  <div
    className="sc-dlniIP EYkDq"
  >
    <div
      className="sc-gtsqUy sc-hKFymg eucOOU jwwTFT"
      columns="10% 35% 30% 15% 10%"
    >
      <div>
        1
      </div>
      <div>
        Nezumi
      </div>
      <div>
        123
      </div>
      <div />
    </div>
    <div
      className="sc-gtsqUy sc-hKFymg eucOOU jwwTFT"
      columns="10% 35% 30% 15% 10%"
    >
      <div>
        2
      </div>
      <div>
        Loi Phan
      </div>
      <div>
        453
      </div>
      <div />
    </div>
  </div>
</div>
`;
